pkg/plugins/runtime/gateway/listener_generator.go

```diff
@@ -1,21 +1,243 @@
+	gw := findMatchingGateway(l.Resources, proxy.Dataplane)
+	if gw == nil {
+		log.V(1).Info("no matching gateway for dataplane",
+			"name", proxy.Dataplane.Meta.GetName(),
+			"mesh", proxy.Dataplane.Meta.GetMesh(),
+		)
+
+		return nil, nil
+	}
+
+	log.V(1).Info(fmt.Sprintf("matched gateway %q to dataplane %q",
+		gw.Meta.GetName(), proxy.Dataplane.Meta.GetName()))
+
+	// Multiple listener specifications can have the same port. If
+	// they are compatible, then we can collapse those specifications
+	// down to a single listener.
+	collapsed := map[uint32][]*mesh_proto.Gateway_Listener{}
+	for _, ep := range gw.Spec.GetConf().GetListeners() {
+		collapsed[ep.GetPort()] = append(collapsed[ep.GetPort()], ep)
+	}
+
+	// A Gateway is a single service across all listeners.
+	service := proxy.Dataplane.Spec.GetIdentifyingService()
+
+	resources := core_xds.NewResourceSet()
+
+	// Generate a listener resource for each port.
+	for port, listeners := range collapsed {
+		protocol := listeners[0].GetProtocol()
+		address := proxy.Dataplane.Spec.GetNetworking().Address
+
+		log.V(1).Info("generating listener",
+			"address", address,
+			"port", port,
+			"protocol", protocol,
+		)
+
+		// TODO(jpeach) verify that the listeners are compatible.
+		// TODO(jpeach) hoist the compatibility check and use it in Gateway validation.
+
+		// This check forces all listeners on the port to have
+		// the same protocol, which is unnecessarily strict. We
+		// cal allow TLS and HTTPS on the same port, for example.
+		for i := range listeners {
+			if listeners[i].GetProtocol() != listeners[0].GetProtocol() {
+				return nil, errors.Errorf("cannot collapse listener protocols %s and %s",
+					listeners[i].GetProtocol(), listeners[0].GetProtocol(),
+				)
+			}
+		}
+
+		switch protocol {
+		case mesh_proto.Gateway_Listener_UDP:
+			fallthrough
+		case mesh_proto.Gateway_Listener_TCP:
+			fallthrough
+		case mesh_proto.Gateway_Listener_TLS:
+			fallthrough
+		case mesh_proto.Gateway_Listener_HTTPS:
+			return nil, errors.Errorf("unsupported protocol %q", protocol)
+		}
+
+		filters := envoy_listeners.NewFilterChainBuilder(proxy.APIVersion)
+
+		// For HTTP protocols, add the connection manager and
+		// (for now), send all traffic to the default route,
+		// just to keep the xDS snapshot consistent.
+		switch protocol {
+		case mesh_proto.Gateway_Listener_HTTP:
+			filters.Configure(
+				envoy_listeners.HttpConnectionManager(service, false),
+				envoy_listeners.HttpDynamicRoute(DefaultRouteName),
+			)
+		case mesh_proto.Gateway_Listener_HTTPS:
+			filters.Configure(
+				envoy_listeners.HttpConnectionManager(service, true),
+				envoy_listeners.HttpDynamicRoute(DefaultRouteName),
+			)
+		}
+
+		// Add edge proxy recommendations.
+		filters.Configure(
+			envoy_listeners.AddFilterChainConfigurer(
+				v3.HttpConnectionManagerMustConfigureFunc(func(hcm *envoy_hcm.HttpConnectionManager) {
```

Comment:
will content of this closoure be refactored to configurers?

I think the rule of thumb we may apply is that we are trying to avoid `github.com/envoyproxy/go-control-plane/` imports in generators.

1) Generators are usually very "busy". In initial version of Kuma we were managing Envoy config in generators. I think that refactoring to configurers and abstracting away Envoy config might have been one of the best technical decision in this project.
2) Envoy V4 will come eventually, configurers abstraction helped us A LOT with V2->V3 migration.

Reply:
It's not clear whether there will ever be a v4, though certainly the protobuf API will evolve.

Maybe we can refactor this out into the configurers, I'm just a bit concerned about the about of boilerplate and indirection involved in that.

Reply:
I think the initial plan was to have a new API version every year but they underestimated how much effort is to migrate everything across ecosystem. But I still think the V4 will come eventually, they have to break the API at some point.